src/garage/torch/algos/vpg.py

```diff
@@ -135,44 +141,80 @@ def train_once(self, itr, paths):
+            for start in range(0, len(rewards_flat), step_size):
+                ids = shuffled_ids[start:start + step_size]
+                loss = self._train(obs_flat[ids], actions_flat[ids],
+                                   rewards_flat[ids], advantages_flat[ids])
+            logger.log('Epoch: {} | Loss: {}'.format(epoch, loss))
 
-        kl_before = self._compute_kl_constraint(obs).detach()
-        self._optimize(itr, obs, actions, rewards, valids, baselines)
+        self.baseline.fit(paths)
 
         with torch.no_grad():
-            loss_after = self._compute_loss(itr, obs, actions, rewards, valids,
-                                            baselines)
-            kl = self._compute_kl_constraint(obs)
-            policy_entropy = self._compute_policy_entropy(obs)
-
-        average_returns = log_performance(itr,
-                                          TrajectoryBatch.from_trajectory_list(
-                                              self.env_spec, paths),
-                                          discount=self.discount)
+            loss_after = self._compute_loss(obs_flat, actions_flat,
+                                            rewards_flat, advantages_flat)
+            kl_after = self.compute_kl_constraint(obs_flat)
+            policy_entropy = self.compute_policy_entropy(obs_flat)
 
         with tabular.prefix(self.policy.name):
-            tabular.record('LossBefore', loss.item())
-            tabular.record('LossAfter', loss_after.item())
-            tabular.record('dLoss', loss.item() - loss_after.item())
-            tabular.record('KLBefore', kl_before.item())
-            tabular.record('KL', kl.item())
-            tabular.record('Entropy', policy_entropy.mean().item())
+            tabular.record('/LossBefore', loss_before.item())
+            tabular.record('/LossAfter', loss_after.item())
+            tabular.record('/dLoss', loss_before.item() - loss_after.item())
+            tabular.record('/KLBefore', kl_before.item())
+            tabular.record('/KL', kl_after.item())
+            tabular.record('/Entropy', policy_entropy.mean().item())
 
-        self.baseline.fit(paths)
-        return np.mean(average_returns)
+        self._old_policy.load_state_dict(self.policy.state_dict())
+
+        undiscounted_returns = log_performance(
+            itr,
+            TrajectoryBatch.from_trajectory_list(self.env_spec, paths),
+            discount=self.discount)
+        return np.mean(undiscounted_returns)
+
+    def _train(self, obs, actions, rewards, advantages):
+        """Train the algorithm with minibatch.
 
-    def _compute_loss(self, itr, obs, actions, rewards, valids, baselines):
+        Args:
+            obs (torch.Tensor): Observation from the environment.
+            actions (torch.Tensor): Predicted action.
+            rewards (torch.Tensor): Feedback from the environment.
+            advantages (torch.Tensor): Expected rewards over the actions.
```

Comment:
advantages are not the expected rewards. they are the difference between the expected returns and the predicted returns.

Reply:
fixed